assertsprocessor/span_enrichment.go

```diff
@@ -15,67 +17,108 @@ const (
+}
+
+func buildCompiledConfig(logger *zap.Logger, config *Config) (map[string]map[string][]*customAttributeConfigCompiled, error) {
+	compiledAttributes := map[string]map[string][]*customAttributeConfigCompiled{}
+	for targetAtt, attrConfigsByServiceKey := range config.CustomAttributeConfigs {
+		compiledAttributes[targetAtt] = map[string][]*customAttributeConfigCompiled{}
+		for serviceKey, attrConfigs := range attrConfigsByServiceKey {
+			if compiledAttributes[targetAtt][serviceKey] == nil {
+				compiledAttributes[targetAtt][serviceKey] = make([]*customAttributeConfigCompiled, 0)
+			}
+			// Make one pass to ensure all configurations are valid
+			for _, attrConfig := range attrConfigs {
+				err := attrConfig.validate(targetAtt, serviceKey)
+				if err != nil {
+					return nil, err
+				}
+			}
+			for _, attrConfig := range attrConfigs {
+				logger.Debug("Added custom attribute for ",
+					zap.String("TargetAttribute", targetAtt),
+					zap.String("Service Key", serviceKey),
+					zap.String("Source Attributes", "["+strings.Join(attrConfig.SourceAttributes, ", ")+"]"),
+					zap.String("Regex", attrConfig.RegExp),
+					zap.String("Replacement", attrConfig.Replacement))
+				compiledAttributes[targetAtt][serviceKey] = append(compiledAttributes[targetAtt][serviceKey], attrConfig.compile())
+			}
 		}
 	}
-	return &processor
+	return compiledAttributes, nil
+}
+
+// configListener interface implementation
+func (ep *spanEnrichmentProcessorImpl) isUpdated(currConfig *Config, newConfig *Config) bool {
+	updated := !reflect.DeepEqual(currConfig.CustomAttributeConfigs, newConfig.CustomAttributeConfigs)
+	if updated {
+		ep.logger.Info("Change detected in config CustomAttributeConfigs",
+			zap.Any("Current", currConfig.CustomAttributeConfigs),
+			zap.Any("New", newConfig.CustomAttributeConfigs),
+		)
+	} else {
+		ep.logger.Debug("No change detected in config CustomAttributeConfigs")
+	}
+	return updated
+}
+
+func (ep *spanEnrichmentProcessorImpl) onUpdate(newConfig *Config) error {
+	newAttributes, err := buildCompiledConfig(ep.logger, newConfig)
+	if err == nil {
+		ep.logger.Info("Updated config RequestContextExps",
+			zap.Any("New", newConfig.CustomAttributeConfigs),
+		)
+		ep.configRWMutex.Lock()
+		ep.customAttributes = newAttributes
+		ep.configRWMutex.Unlock()
+	} else {
+		ep.logger.Error("Ignoring config RequestContextExps due to regex compilation error", zap.Error(err))
+	}
+	return err
 }
 
 func (ep *spanEnrichmentProcessorImpl) enrichSpan(namespace string, service string, span *ptrace.Span) {
 	ep.addRequestType(span)
-	ep.addRequestContext(namespace, service, span)
-	ep.addErrorType(span)
+	ep.configRWMutex.RLock()
+	currentConfig := ep.customAttributes
+	ep.configRWMutex.RUnlock()
+	for targetAtt, configByServiceKey := range currentConfig {
+		if configByServiceKey[namespace+"#"+service] != nil {
+			for _, config := range configByServiceKey[namespace+"#"+service] {
```

Comment:
If the service specific config don't find a value, we should use the default. `addCustomAttribute` can be modified to `getCustomAttribute`. This way we can first use the service specific configs and in case of no match use the default config